packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts

```diff
@@ -0,0 +1,241 @@
+    }
+
+    function isParameterPropertyWithName(
+      node: TSESTree.Parameter,
+      name: string,
+    ): boolean {
+      return (
+        node.type === AST_NODE_TYPES.TSParameterProperty &&
+        ((node.parameter.type === AST_NODE_TYPES.Identifier && // constructor (public foo) {}
+          node.parameter.name === name) ||
+          (node.parameter.type === AST_NODE_TYPES.AssignmentPattern && // constructor (public foo = 1) {}
+            node.parameter.left.type === AST_NODE_TYPES.Identifier &&
+            node.parameter.left.name === name))
+      );
+    }
+
+    function getIdentifier(node: TSESTree.Node): TSESTree.Identifier | null {
+      if (node.type === AST_NODE_TYPES.Identifier) {
+        return node;
+      }
+      if (
+        node.type === AST_NODE_TYPES.TSAsExpression ||
+        node.type === AST_NODE_TYPES.TSNonNullExpression
+      ) {
+        return getIdentifier(node.expression);
+      }
+      return null;
+    }
+
+    return {
+      ClassBody(): void {
+        reportInfoStack.push({
+          unnecessaryAssigments: [],
+          assignedBeforeUnnecessary: new Set(),
+          assignedBeforeCtor: new Set(),
+        });
+      },
+      'ClassBody:exit'(): void {
+        const { unnecessaryAssigments, assignedBeforeCtor } = nullThrows(
+          reportInfoStack.pop(),
+          'The top stack should exist',
+        );
+        unnecessaryAssigments.forEach(({ name, node }) => {
+          if (assignedBeforeCtor.has(name)) {
+            return;
+          }
+          context.report({
+            node,
+            messageId: 'unnecessaryAssign',
+          });
+        });
+      },
+      'PropertyDefinition AssignmentExpression'(
+        node: TSESTree.AssignmentExpression,
+      ): void {
+        if (!isThisMemberExpression(node.left)) {
+          return;
+        }
+
+        const name = getPropertyName(node.left);
+
+        if (!name) {
+          return;
+        }
+
+        const functionNode = findFunction(node);
+        if (functionNode) {
+          const isArrowIIFE =
+            functionNode.type === AST_NODE_TYPES.ArrowFunctionExpression &&
+            functionNode.parent.type === AST_NODE_TYPES.CallExpression;
+
+          if (
+            !(
+              isArrowIIFE &&
+              findPropertyDefinition(node)?.value === functionNode.parent
+            )
+          ) {
+            return;
+          }
+        }
+
+        const { assignedBeforeCtor } = nullThrows(
+          reportInfoStack.at(reportInfoStack.length - 1),
+          'The top stack should exist',
+        );
+        assignedBeforeCtor.add(name);
+      },
+      "MethodDefinition[kind='constructor'] > FunctionExpression AssignmentExpression"(
+        node: TSESTree.AssignmentExpression,
+      ): void {
+        if (!isThisMemberExpression(node.left)) {
+          return;
+        }
+
+        const leftName = getPropertyName(node.left);
```

Comment:
[Refactor] Both `getPropertyName` calls are preceded by the `if (!isThisMemberExpression(node.left)) {`. Moving the `isThisMemberExpression` check inside `getPropertyName` would clean up the code a bit, I think.

```suggestion
        const leftName = getPropertyName(node.left);
```